src/ui/mapruler.cpp

```diff
@@ -0,0 +1,137 @@
+#include "mapruler.h"
+#include "metatile.h"
+
+#include <QGraphicsItem>
+#include <QGraphicsSceneMouseEvent>
+#include <QPainter>
+#include <QColor>
+#include <QToolTip>
+
+int MapRuler::thickness = 3;
+
+
+void MapRuler::init() {
+    setVisible(false);
+    setPoints(QPoint(), QPoint());
+    anchored = false;
+    locked = false;
+    statusMessage = QString("Ruler: 0");
+    xRuler = QRect();
+    yRuler = QRect();
+    cornerTick = QLine();
+}
+
+QRectF MapRuler::boundingRect() const {
+    return QRectF(-thickness, -thickness, pixWidth() + thickness * 2, pixHeight() + thickness * 2);
+}
+
+QPainterPath MapRuler::shape() const {
+    QPainterPath ruler;
+    ruler.setFillRule(Qt::WindingFill);
+    ruler.addRect(xRuler);
+    ruler.addRect(yRuler);
+    ruler = ruler.simplified();
+    for (int x = 17.5; x < pixWidth(); x += 16)
+        ruler.addRect(x, xRuler.y(), 0, thickness);
+    for (int y = 17.5; y < pixHeight(); y += 16)
+        ruler.addRect(yRuler.x(), y, thickness, 0);
+    return ruler;
+}
+
+void MapRuler::paint(QPainter *painter, const QStyleOptionGraphicsItem *, QWidget *) {
+    painter->setPen(QPen(borderColor));
+    painter->setBrush(QBrush(innerColor));
+    painter->drawPath(shape());
+    if (deltaX() && deltaY())
+        painter->drawLine(cornerTick);
+}
+
+void MapRuler::setAnchor(const QPointF &scenePos, const QPoint &screenPos) {
+    anchored = true;
+    locked = false;
+    QPoint tilePos = Metatile::coordFromPixmapCoord(scenePos);
+    setPoints(tilePos, tilePos);
+    updateGeometry();
+    setVisible(true);
+    showDimensions(screenPos);
+}
+
+void MapRuler::endAnchor() {
+    hideDimensions();
+    prepareGeometryChange();
+    init();
+}
+
+void MapRuler::setEndPos(const QPointF &scenePos, const QPoint &screenPos) {
+    if (locked)
+        return;
+    QPoint pos = Metatile::coordFromPixmapCoord(scenePos);
+    QPoint lastEndPos = endPos();
+    setP2(pos);
+    if (pos != lastEndPos)
+        updateGeometry();
+    showDimensions(screenPos);
+}
+
+bool MapRuler::isMousePressed(QGraphicsSceneMouseEvent *event) const {
+    return event->buttons() & acceptedMouseButtons() && event->type() == QEvent::GraphicsSceneMousePress;
+}
+
+void MapRuler::showDimensions(const QPoint &screenPos) {
+    // This is a hack to make the tool tip follow the cursor since it won't change position if the text is the same.
+    QToolTip::showText(screenPos + QPoint(16, -8), statusMessage + ' ');
+    QToolTip::showText(screenPos + QPoint(16, -8), statusMessage);
```

Comment:
It would be nice if the tooltip stayed visible for more than one second.  `QToolTip::showText` takes a "duration" parameter, but I couldn't get it to work.

Reply:
Yeah, I played around with that and couldn't get it to work either. Earlier in this branch I was rendering the dimensions as a part of the ruler graphic using `QPainter::drawText()` but the problem with that approach was that the size of the text was fixed and was hard to see when zoomed out, and was hard to see against dark map tiles. For those reasons it just seemed cleaner and easier to use a tooltip. I'll play around with this more and try to find a better solution.